### PR TITLE
fake bucket implementation fix for HNS APIs

### PR DESCRIPTION
**Pull request by Tulsishah**
### Description
- **CreateFolder**: In the hierarchical bucket, control client API creates folders  as well as prefixes for backward compatibility. Therefore, a prefix object entry needs to be created.
- **DeleteFolder**: In the hierarchical bucket, control client API deletes folders as well as prefixes for backward compatibility. Therefore, a prefix object entry needs to be deleted.
- **ListObjects**: To include created folder only as prefix in the list, added check.
- **RenameFolder**: Change all the entries with given source prefix to destination prefix.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Automated
3. Integration tests - NA

### CODE REVIEW

File: internal/storage/fake/bucket.go [hunk 1]
```diff
@@ -973,6 +1004,25 @@ func (b *bucket) CreateFolder(ctx context.Context, folderName string) (*gcs.Fold
...
+	// Find any existing record for this name.
+	existingObjectPrefixIndex := b.objects.find(folderName)
+
+	// Create a prefix object record from the given attributes.
+	var prefixObject fakeObject
+	prefixObject.metadata = gcs.Object{Name: folderName}
+
+	// Replace an entry in or add an entry to our list of objects.
+	if existingObjectPrefixIndex < len(b.objects) {
+		b.objects[existingIndex] = prefixObject
```
Comment (vadlakondaswetha) [inline internal/storage/fake/bucket.go:1020]: are you supposed to use existingObjectPrefixIndex here?

File: internal/storage/fake/bucket.go [hunk 2]
```diff
@@ -992,29 +1042,33 @@ func (b *bucket) RenameFolder(ctx context.Context, folderName string, destinatio
 		return nil, err
 	}
 
-	dst := b.folders[srcIndex]
-	dst.Name = destinationFolderId
+	// Find all folders starting with the given prefix and update their names.
+	for i := range b.folders {
+		if strings.HasPrefix(b.folders[i].Name, folderName) {
```
Comment (vadlakondaswetha) [inline internal/storage/fake/bucket.go:1047]: Rename folder is for a specific folder right. why are you doing this prefix check?

Comment (Tulsishah) [reply internal/storage/fake/bucket.go:1047]: Let's say we have the following structure:
```
A/B
A/a.txt
A/C/d.txt
```
If we rename "A" to "A_new", we need to check for the prefix "A/" in both the folder and object arrays to rename "A/B" to "A_new/B" and "A/a.txt" to "A_new/a.txt".

File: internal/storage/fake/bucket.go (hunk 1 above)
Comment (Tulsishah) [reply internal/storage/fake/bucket.go]: Done thanks.

File: internal/storage/fake/bucket.go (hunk 2 above)
Comment (vadlakondaswetha) [reply internal/storage/fake/bucket.go:1047]: Got it. It is to update the child directories.

**Review by vadlakondaswetha — APPROVED** [review]

**Pull request merged**